File: data-structures-and-algorithms-ashrafi/java/src/recursive/problem17/Problem17.java
public class Problem17 {
    
    public static double calculetForX(String s, int x){

        s = s.replace(" ", "");

        if (s.equals("x")) {
            return x;
        } else {
            for (int i = s.length() - 1; i >= 0; i--) {
                char operator = s.charAt(i);
                if (operator == '+'){
                  return calculetForX(s.substring(0, i), x) + calculetForX(s.substring(i + 1), x);   
                }
                if (operator == '-'){
                     return calculetForX(s.substring(0, i), x) - calculetForX(s.substring(i + 1), x);
                }
            }
            for (int i = 0; i < s.length(); i++) {
                char operator = s.charAt(i);
                if (operator == '*'){
                     return calculetForX(s.substring(0, i), x) * calculetForX(s.substring(i + 1), x);
                }
                if (operator == '/'){
                     return calculetForX(s.substring(0, i), x) / calculetForX(s.substring(i + 1), x);
                }
            }
        }
        return Double.parseDouble(s);
    }

    public static void main(String[] args) {
        String s = "x + x * x / x - x + x + x - x + x";
        int x = 2;

        System.out.printf("Answer is = %.2f%n", calculetForX(s, x));
    }
}

File: data-structures-and-algorithms-ashrafi/python/src/recursive/problem_17/problem_17.py
def calculet_for_x(s, x):
    
    s = s.replace(" ", "")
    
    if s == "x":
        return x
    else:
        for i in range(len(s) - 1, 0, -1):
            c = s[i]
            if c == "+":
                return calculet_for_x(s[0:i], x) + calculet_for_x(s[i + 1:], x)
            elif c == "-":
                return calculet_for_x(s[0:i], x) - calculet_for_x(s[i + 1:], x)        
        for i in range(len(s) - 1, 0, -1):
            c = s[i]
            if c == "*":
                return calculet_for_x(s[0:i], x) * calculet_for_x(s[i + 1:], x)
            elif c == "/":
                return calculet_for_x(s[0:i], x) / calculet_for_x(s[i + 1:], x)
          
        return float(s)      

def main():
    s = "x + x * x / x - x + x + x - x + x"
    x = 2
    
    result = calculet_for_x(s, x)
    print(f"Result = {result}")


if __name__ == "__main__":
    main()